subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c

```diff
@@ -0,0 +1,1303 @@
+#define TOPIC_EVENTS		"devices/%s/messages/events/"
+#define TOPIC_TWIN_REQUEST	"$iothub/twin/GET/?$rid=%d"
+#define TOPIC_DIRECT_METHODS	"$iothub/methods/POST/#"
+#define TOPIC_DIRECT_METHOD_RES	"$iothub/methods/res/%d/?$rid=%d"
+
+#define DPS_USER_NAME		CONFIG_AZURE_IOT_HUB_DPS_ID_SCOPE \
+				"/registrations/%s/api-version=2019-03-31"
+
+#if IS_ENABLED(CONFIG_AZURE_IOT_HUB_DEVICE_ID_APP)
+static char device_id_buf[CONFIG_AZURE_IOT_HUB_DEVICE_ID_MAX_LEN + 1];
+#else
+static char device_id_buf[CONFIG_AZURE_IOT_HUB_DEVICE_ID_MAX_LEN + 1] =
+	CONFIG_AZURE_IOT_HUB_DEVICE_ID;
+#endif
+
+static struct azure_iot_hub_config conn_config = {
+	.device_id = device_id_buf,
+#if IS_ENABLED(CONFIG_AZURE_IOT_HUB_DEVICE_ID_APP)
+	.device_id_len = 0,
+#else
+	.device_id_len = sizeof(device_id_buf) - 1,
+#endif
+};
+
+static azure_iot_hub_evt_handler_t evt_handler;
+
+/* If DPS is used, the IoT Hub hostname is obtained through that service,
+ * otherwise it has to be set compile time using CONFIG_AZURE_IOT_HUB_HOSTNAME.
+ * The maximal size is length of hub name + device ID length + lengths of
+ * ".azure-devices-provisioning.net/" and "/?api-version=2018-06-30"
+ */
+#if IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS)
+#define USER_NAME_BUF_LEN	(CONFIG_AZURE_IOT_HUB_HOSTNAME_MAX_LEN + \
+				CONFIG_AZURE_IOT_HUB_DEVICE_ID_MAX_LEN + \
+				sizeof(".azure-devices-provisioning.net/") + \
+				sizeof("/?api-version=2018-06-30"))
+static char user_name_buf[USER_NAME_BUF_LEN];
+static struct mqtt_utf8 user_name = {
+	.utf8 = user_name_buf,
+};
+#else
+static struct mqtt_utf8 user_name = {
+	.utf8 = USER_NAME_STATIC,
+	.size = sizeof(USER_NAME_STATIC) - 1,
+};
+#endif /* IS_ENABLED(CONFIG_AZURE_IOT_HUB_DPS) */
+
+enum connection_state {
+	STATE_IDLE,
+	STATE_INIT,
+	STATE_CONNECTING,
+	STATE_CONNECTED,
+	STATE_DISCONNECTING,
+};
+
+static enum connection_state connection_state = STATE_IDLE;
+static char rx_buffer[CONFIG_AZURE_IOT_HUB_MQTT_RX_TX_BUFFER_LEN];
+static char tx_buffer[CONFIG_AZURE_IOT_HUB_MQTT_RX_TX_BUFFER_LEN];
+static char payload_buf[CONFIG_AZURE_IOT_HUB_MQTT_PAYLOAD_BUFFER_LEN];
+
+static bool is_initialized;
+static struct mqtt_client client;
+static struct sockaddr_storage broker;
+static K_SEM_DEFINE(connection_poll_sem, 0, 1);
+static K_SEM_DEFINE(connected, 0, 1);
+static atomic_t dps_disconnecting;
+
+/* Build time asserts */
+
+/* Device ID source must be one of three for the library to work:
+ *	- Kconfigured IoT hub device ID: CONFIG_AZURE_IOT_HUB_DEVICE_ID
+ *	- Runtime app-provided device ID
+ */
+BUILD_ASSERT((sizeof(CONFIG_AZURE_IOT_HUB_DEVICE_ID) - 1 > 0) ||
+	     IS_ENABLED(CONFIG_AZURE_IOT_HUB_DEVICE_ID_APP),
+	     "Device ID must be set by Kconfig or application");
+
+/* Static function signatures */
+static int connect_client(struct azure_iot_hub_config *cfg);
+
+/* Static functions */
+static void azure_iot_hub_notify_event(struct azure_iot_hub_evt *evt)
+{
+	evt_handler(evt);
```

Comment:
Probably handled somewhere else in the library, but you could add a check here to make sure the handler is not null.

Reply:
It should not be possible for events to happen if `evt_handler == NULL`, but I'll add a check anyway